ocaml/typing/typecore.ml

```diff
@@ -6534,9 +6534,7 @@ and type_ident env ?(recarg=Rejected) lid =
        (* if the locality of returned value of the primitive is poly
           we then register allocation for further optimization *)
        | (Prim_poly, _), Some mode ->
-           register_allocation_mode
-             (Alloc.meet [Alloc.max_with (Comonadic Areality) mode;
-                          Alloc.max_with (Comonadic Linearity) Linearity.many])
```

Comment:
It's not obvious to me this is the same?  What happened to the linearity constraint?

Reply:
They are almost the same. The logic here is to handle `[@local_opt]` in primitives, and the `mode` is the locality mode instantiated to the current call site. We just want to register `mode` here (to be potentially pushed to `local`). However, the currently interface of `register_allocation_mode` is bloated and requires all other axes as well. So here we give the other axes as well but fixed (and therefore won't be touched by `optimise_allocations`).

The old code assumes that external are legacy in all other axes, which is wrong. Consider:
```
external foo : 'a -> 'a @ once [@local_opt]
```
where the return mode is polymorphic in locality but has non-legacy `once` in `linearity`. But even in this case, `once` is fixed so doesn't need to be registered.

Reply:
(Zesen explained a little more out of band.  My understanding is: what the old code was doing is unnecessary, because `register_allocation_mode` has no effect on axes that are constant, and the linearity axis is constant either way.  The new version just skips that extra work.)